### PR TITLE
[MM-28672] Add e2e testing for user invite 

### PR DESCRIPTION
**Pull request by M-Buntoro**
#### Summary
Add end to end test case for `mmctl user invite` 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-28672
Fixes https://github.com/mattermost/mattermost-server/issues/15487

### CODE REVIEW

File: commands/user_e2e_test.go [hunk 1]
```diff
@@ -136,3 +136,59 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
 }
+
+func (s *MmctlE2ETestSuite) TestUserInviteCmdf() {
+	s.SetupTestHelper().InitBasic()
+
+	s.RunForSystemAdminAndLocal("Invite user", func(c client.Client) {
+
+		s.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableEmailInvitations = true })
+		defer func() {
+			s.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableEmailInvitations = false })
```
Comment (isacikgoz) [inline commands/user_e2e_test.go:147]: Nit: I'd prefer to restore the config with the existing value. We can't be sure if it is going to be always `false`.

File: commands/user_e2e_test.go [hunk 2]
```diff
@@ -136,3 +136,59 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+
+		err := userInviteCmdF(c, &cobra.Command{}, []string{s.th.BasicUser.Email, s.th.BasicTeam.Id})
+
+		s.Require().Nil(err)
+		s.Require().Len(printer.GetLines(), 1)
+		s.Require().Equal(printer.GetLines()[0], "Invites may or may not have been sent.")
+		s.Require().Len(printer.GetErrorLines(), 0)
+	})
+
+	s.RunForSystemAdminAndLocal("Inviting when email invitation disabled", func(c client.Client) {
```
Comment (isacikgoz) [inline commands/user_e2e_test.go:160]: Same here, let's be sure the setting is disabled.

File: commands/user_e2e_test.go [hunk 3]
```diff
@@ -136,3 +136,59 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+		s.Require().Len(printer.GetLines(), 0)
+		s.Require().Len(printer.GetErrorLines(), 1)
+		s.Require().Equal(printer.GetErrorLines()[0], "Unable to invite user with email "+s.th.BasicUser.Email+" to team "+s.th.BasicTeam.Name+". Error: : Email invitations are disabled., ")
+	})
+
+	s.RunForSystemAdminAndLocal("Invite user not accepted domain", func(c client.Client) {
+
+		s.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableEmailInvitations = true })
+		defer func() {
+			s.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableEmailInvitations = false })
```
Comment (isacikgoz) [inline commands/user_e2e_test.go:175]: Same as above :)

File: commands/user_e2e_test.go [hunk 4]
```diff
@@ -136,3 +136,59 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
 		s.Equal("Unable to find user '"+emailArg+"'", printer.GetErrorLines()[0])
 	})
 }
+
+func (s *MmctlE2ETestSuite) TestUserInviteCmdf() {
+	s.SetupTestHelper().InitBasic()
+
+	s.RunForSystemAdminAndLocal("Invite user", func(c client.Client) {
+
```
Comment (isacikgoz) [inline commands/user_e2e_test.go:144]: FYI: linter complains about these unnecessary leading newlines after the function declarations.
```suggestion
```

File: commands/user_e2e_test.go [hunk 5]
```diff
@@ -136,3 +136,59 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+	s.RunForSystemAdminAndLocal("Inviting when email invitation disabled", func(c client.Client) {
+
+		printer.Clean()
+		err := userInviteCmdF(c, &cobra.Command{}, []string{s.th.BasicUser.Email, s.th.BasicTeam.Id})
+
+		s.Require().Nil(err)
+		s.Require().Len(printer.GetLines(), 0)
+		s.Require().Len(printer.GetErrorLines(), 1)
+		s.Require().Equal(printer.GetErrorLines()[0], "Unable to invite user with email "+s.th.BasicUser.Email+" to team "+s.th.BasicTeam.Name+". Error: : Email invitations are disabled., ")
```
Comment (isacikgoz) [inline commands/user_e2e_test.go:168]: We can use `fmt.Sprintf()` instead of concatenating with variables.

**Review by isacikgoz — CHANGES_REQUESTED** [review]
Thanks for the contribution @M-Buntoro!! 🚀 I'd like request some minor changes to improve code readability etc.

File: commands/user_e2e_test.go [hunk 6]
```diff
@@ -136,3 +136,59 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+	s.SetupTestHelper().InitBasic()
+
+	s.RunForSystemAdminAndLocal("Invite user", func(c client.Client) {
+
+		s.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableEmailInvitations = true })
+		defer func() {
+			s.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableEmailInvitations = false })
+		}()
+
+		printer.Clean()
```
Comment (mgdelacroix) [inline commands/user_e2e_test.go:150]: nit: for consistency, please make this the first line of the test case for all test cases

**Review by mgdelacroix — COMMENTED** [review]
Thanks for your contribution @M-Buntoro!! Looking good!

Added one comment besides the ones from @isacikgoz. Would you be open to add a final test case to cover when an unprivileged user (`s.th.BasicUser` for example) runs the command?

File: commands/user_e2e_test.go (hunk 1 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:147]: Ah good point, let me adjust that

File: commands/user_e2e_test.go (hunk 2 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:160]: Yes, agree

File: commands/user_e2e_test.go (hunk 6 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:150]: Sure, I'll make sure it is the first on all test case

File: commands/user_e2e_test.go (hunk 5 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:168]: Sure

File: commands/user_e2e_test.go (hunk 4 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:144]: noted, will remove the space

File: commands/user_e2e_test.go (hunk 3 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:175]: yes

**Comment by M-Buntoro** [comment]
Sure @mgdelacroix I'll add the case for that :ok_hand:

File: commands/user_e2e_test.go [hunk 7]
```diff
@@ -136,3 +138,67 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+		team.AllowedDomains = "@example.com"
+		team, appErr := s.th.App.UpdateTeam(team)
+		s.Require().Nil(appErr)
+
+		user := s.th.CreateUser()
+		err := userInviteCmdF(c, &cobra.Command{}, []string{user.Email, team.Id})
+		s.Require().Nil(err)
+		s.Require().Len(printer.GetLines(), 0)
+		s.Require().Len(printer.GetErrorLines(), 1)
+		s.Require().Equal(printer.GetErrorLines()[0], "Unable to invite user with email "+user.Email+" to team "+team.Name+". Error: : The following email addresses do not belong to an accepted domain: "+user.Email+". Please contact your System Administrator for details., ")
```
Comment (isacikgoz) [inline commands/user_e2e_test.go:202]: We can use `fmt.Sprintf` here.

Comment (M-Buntoro) [reply commands/user_e2e_test.go:202]: ah sorry, missed that one, okay

File: commands/user_e2e_test.go [hunk 8]
```diff
@@ -136,3 +138,72 @@ func (s *MmctlE2ETestSuite) TestSearchUserCmd() {
...
+	s.SetupTestHelper().InitBasic()
+
+	s.RunForAllClients("Invite user", func(c client.Client) {
+		printer.Clean()
+
+		previousVal := s.th.App.Config().ServiceSettings.EnableEmailInvitations
+		s.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableEmailInvitations = true })
+		defer func() {
+			s.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableEmailInvitations = *previousVal })
+		}()
```
Comment (mgdelacroix) [inline commands/user_e2e_test.go:152]: nit: nonblocking, but being the defer a oneliner, there is no need to wrap it in a `func`, same with the others:

```suggestion
		defer s.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableEmailInvitations = *previousVal })
```

**Review by mgdelacroix — APPROVED** [review]
Thanks for your contribution @M-Buntoro, looks good to me!! :rocket: 

Added only one nonblocking comment, feel free to make a decision on the change

File: commands/user_e2e_test.go (hunk 8 above)
Comment (M-Buntoro) [reply commands/user_e2e_test.go:152]: cool, its okay on me @mgdelacroix

**Review by isacikgoz — APPROVED** [review]
Awesome!! Thanks for the contribution 🎉

**Review by lindy65 — APPROVED** [review]
Thanks @M-Buntoro @isacikgoz - looks good to me 👍

**Comment by hanzei** [comment]
Another one! Thanks again @M-Buntoro :rocket:

**Pull request merged**